### PR TITLE
Dispose the ContentFileProvider when the host is disposed

### LINKED ISSUE #36550: Properly dispose the PhysicalFileProvider when Host is disposed
Today the PhysicalFileProvider gets finalized instead of being properly disposed by the Host.

### PR DESCRIPTION
**Pull request by davidfowl**
- Today we don't and it results in a leak when hosts are created and destroyed.

Contributes to https://github.com/dotnet/aspnetcore/issues/31125

Fixed https://github.com/dotnet/runtime/issues/36550

### CODE REVIEW

**Comment by eerhardt** [comment]
Is there a test that can be added for this?

**Comment by davidfowl** [comment]
> Is there a test that can be added for this?

It's good you asked, I plan to but I can't get test working locally 😄. I'll write one then use the CI

**Comment by eerhardt** [comment]
> but I can't get test working locally

😮 , Do either of the following work for you:

* from the repo root: `.\.dotnet\dotnet.exe test .\src\libraries\Microsoft.Extensions.Hosting\tests\UnitTests\`
* from the repo root: `.\build.cmd -vs .\src\libraries\Microsoft.Extensions.Hosting\Microsoft.Extensions.Hosting.sln` and then use Test Explorer to run

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 1]
```diff
@@ -131,14 +142,34 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
...
+            }
+            else
             {
-                case IAsyncDisposable asyncDisposable:
-                    await asyncDisposable.DisposeAsync().ConfigureAwait(false);
-                    break;
-                case IDisposable disposable:
-                    disposable.Dispose();
-                    break;
+                // In the rare case that the user replaced the ContentRootFileProvider, dispose it and the one
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:153]: For my education, how can this happen? Setting it on the `HostBuilderContext.HostingEnvironment.ContentRootFileProvider` during a `ConfigureAppConfiguration` callback?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:153]: It's settable 😄

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 2]
```diff
@@ -219,6 +220,8 @@ private void CreateServiceProvider()
             services.AddSingleton<IHost>(_ =>
             {
                 return new Internal.Host(_appServices,
+                    _hostingEnvironment,
+                    _defaultProvider,
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:224]: wouldn't `_defaultProvider` always end up being `_hostingEnvironment.ContentRootFileProvider` here?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:224]: See https://github.com/dotnet/runtime/blob/9785baac4b4404874c964a82c135fa70fee8b087/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L145

**Comment by davidfowl** [comment]
```
Log level is set to Informational (Default).
Loaded 0 test records from the solution cache in 0.607 sec.
Test data store opened in 0.029 sec.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Microsoft.VisualStudio.ProjectSystem.UnconfiguredProjectImpl'.
   at Microsoft.Verify.NotDisposed(IDisposableObservable disposedValue, String message)
   at Microsoft.VisualStudio.ProjectSystem.UnconfiguredProjectImpl.get_LoadedConfiguredProjects()
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.Components.ActiveConfiguredProjectProvider.get_ActiveConfiguredProject()
   at Microsoft.VisualStudio.ProjectSystem.ProjectConfigurationsServiceBase.GetSuggestedProjectConfigurationStep1(Boolean doNotQueryActiveConfiguration)
   at Microsoft.VisualStudio.ProjectSystem.ProjectConfigurationsServiceBase.GetSuggestedProjectConfigurationAsync(Boolean doNotQueryActiveConfiguration)
   at Microsoft.VisualStudio.ProjectSystem.ProjectConfigurationsServiceUtilities.GetSuggestedProjectConfigurationAsync(IProjectConfigurationsService configurationsService)
   at Microsoft.VisualStudio.ProjectSystem.UnconfiguredProjectImpl.<GetSuggestedConfiguredProjectAsync>d__189.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectData.<TryInitCpsProjectOutputPathsAsync>d__61.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectData.<InitCpsProjectOutputPathsAsync>d__60.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectData.<InitProjectOutputPathsAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectData.<BuildProjectDataAsync>d__50.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectSystemProvider.<LoadProjectAsync>d__98.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectSystemProvider.<InternalLoadProjectsAsync>d__93.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectSystemProvider.<<OnSolutionLoaded>b__111_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Extensibility.TaskExtensions.<WithExceptionLogger>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

**Review by eerhardt — APPROVED** [review]
LGTM

**Comment by davidfowl** [comment]
I'm like 99% sure these failures are unrelated 😄

**Review by maryamariyan — APPROVED** [review]

**Pull request merged**